src/mrpro/operators/_ConstraintsOp.py

```diff
@@ -0,0 +1,145 @@
+        if beta_softplus <= 0:
+            raise ValueError(f'parameter beta_softplus must be greater than zero; given {beta_softplus}')
+
+        self.beta_sigmoid = beta_sigmoid
+        self.beta_softplus = beta_softplus
+
+        self.lower_bounds = [bound[0] for bound in bounds]
+        self.upper_bounds = [bound[1] for bound in bounds]
+
+        for lb, ub in bounds:
+            if (ub is not None and lb is not None) and lb > ub:
+                raise ValueError(
+                    f'bounds should be ( (a1,b1), (a2,b2), ...) with ai<=bi if neither ai or bi is None;\
+                        \nbound tuple {lb,ub} is invalid'
+                )
+
+    @staticmethod
+    def sigmoid(x: torch.Tensor, beta: float = 1.0) -> torch.Tensor:
+        """Constraint x to be in the range given by 'bounds'."""
+
+        return F.sigmoid(beta * x)
+
+    @staticmethod
+    def sigmoid_inverse(x: torch.Tensor, beta: float = 1.0) -> torch.Tensor:
+        """Constraint x to be in the range given by 'bounds'."""
+
+        return torch.logit(x) / beta
+
+    @staticmethod
+    def softplus(x: torch.Tensor, beta: int = 1) -> torch.Tensor:
+        """Constrain x to be in (bound,infty)."""
+
+        return F.softplus(x, beta=beta)
+
+    @staticmethod
+    def softplus_inverse(x: torch.Tensor, beta: int = 1) -> torch.Tensor:
+        """Inverse of 'softplus_transformation."""
+
+        return beta * x + torch.log(-torch.expm1(-beta * x))
+
+    def forward(self, x: tuple[torch.Tensor, ...]) -> tuple[torch.Tensor, ...]:
+        """Transform tensors to chosen range.
+
+        Parameters
+        ----------
+        x
+            tensors to be transformed
+
+        Returns
+        -------
+            tensors transformed to the range defined by the chosen bounds
+        """
+        # iterate over the tensors and constrain them if necessary according to the
+        # chosen bounds
+        xc = []
+        for i in range(len(self.lower_bounds)):
+            lb, ub = self.lower_bounds[i], self.upper_bounds[i]
+
+            # distiguish cases
+            if lb is not None and ub is not None:
+                # case (a,b) with a<b and a,b \in R
+                xc.append(lb + (ub - lb) * self.sigmoid(x[i], beta=self.beta_sigmoid))
+
+            elif lb is not None and ub is None:
+                # case (a,None); corresponds to (a, \infty)
+                xc.append(lb + self.softplus(x[i], beta=self.beta_softplus))
+
+            elif lb is None and ub is not None:
+                # case (None,b); corresponds to (-\infty, b)
+                xc.append(ub - self.softplus(-x[i], beta=self.beta_softplus))
+            elif lb is None and ub is None:
+                # case (None,None); corresponds to (-\infty, \infty), i.e. no transformation
+                xc.append(x[i])
+
+        return tuple(xc)
+
+    def inverse(self, xc: tuple[torch.Tensor, ...]) -> tuple[torch.Tensor, ...]:
```

Comment:
This should be *xc
I. E. multiple tensors

Reply:
done